src/riak_dt_hll.erl

```diff
@@ -0,0 +1,376 @@
+%%              Art Cardinality Esitmation Algorithm" from Google, of
+%%              which many of the libs reference.
+%%
+%% Copyright (c) 2016 Basho Technologies, Inc.  All Rights Reserved.
+%%
+%% This file is provided to you under the Apache License,
+%% Version 2.0 (the "License"); you may not use this file
+%% except in compliance with the License.  You may obtain
+%% a copy of the License at
+%%
+%%   http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing,
+%% software distributed under the License is distributed on an
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+%% -------------------------------------------------------------------
+
+-module(riak_dt_hll).
+
+-behaviour(riak_dt).
+
+-include("riak_kv_types.hrl").
+
+-ifdef(EQC).
+-include_lib("eqc/include/eqc.hrl").
+-export([within_error_check/3, gen_op/0, update_expected/3, eqc_state_value/1]).
+-define(NUMTESTS, 1000).
+-endif.
+
+-ifdef(TEST).
+-include_lib("eunit/include/eunit.hrl").
+-endif.
+
+%% API
+-export([new/0, value/1, value/2]).
+-export([update/3, update/4, merge/2, equal/2]).
+-export([to_binary/1, from_binary/1]).
+-export([to_binary/2]).
+-export([stats/1, stat/2]).
+-export([parent_clock/2]).
+-export([to_version/2]).
+
+%% Additional APIs for hyperloglog types
+-export([new/1, reduce/2, precision/1, check_precision_and_reduce/2]).
+
+-export_type([hllset/0, card/0, precision/0]).
+
+-define(HYPER_DEFAULT_BACKEND, hyper_binary).
+-define(STATS, [bytes, precision]).
+
+-define(DT_HLLSET_TAG, 86).
+-define(TAG, ?DT_HLLSET_TAG).
+-define(V1_VERS, 1).
+
+%% Error macros pulled from riak_dt_tags.hrl
+-define(UNSUPPORTED_VERSION(Vers), {error, unsupported_version, Vers}).
+-define(INVALID_BINARY, {error, invalid_binary}).
+-define(UNSUPPORTED_VERSION, {error, unsupported_version,
+                              Vers :: pos_integer()}).
+-define(OPS, [add, add_all]).
+
+-record(hyper, {p :: precision(),
+                registers :: {module(), registers()}}).
+
+-opaque hllset() :: #hyper{}.
+
+-type hllset_op() :: {add|add_all, term()|list(term())} |
+                     {update, [hllset_op()]}.
+
+-type precision() :: hyper:precision().
+-type registers() :: hyper:registers().
+-type card() :: number().
+-type actor() :: riak_dt:actor().
+-type stat() :: bytes | precision.
+
+-spec new() -> hllset().
+new() ->
+    new(?HYPER_DEFAULT_PRECISION).
+
+-spec new(precision()|riak_kv_bucket:props()) -> hllset().
+new(Precision) when is_integer(Precision), Precision > 3
+                    andalso Precision < 17 ->
+    hyper:new(Precision, ?HYPER_DEFAULT_BACKEND);
+new(BProps) when is_list(BProps) ->
+    Precision = proplists:get_value(hll_precision, BProps,
+                                    ?HYPER_DEFAULT_PRECISION),
+    new(Precision);
+new(_) ->
```

Comment:
For this case, I think returning to the default is fine... they want HLL, but maybe don't get precision? I'll def. log/warning w/ it though.